server/aesdsocket.c

```diff
@@ -35,377 +41,643 @@
+    printf("Deleting file\n");
+    unlink("/var/tmp/aesdsocketdata");
+    exit(1);
+}
 
-	// bind it to the port we passed in to getaddrinfo()	
+static void *get_in_addr(struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET) {
+        return &(((struct sockaddr_in*)sa)->sin_addr);
+    }
 
-    	s.sin_addr.s_addr = INADDR_ANY;
-    	s.sin_family = AF_INET;
-    	s.sin_port = htons(PORT);
+    return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
 
-   	status = bind(serverfd , (struct sockaddr_in *)&s , sizeof(struct sockaddr_in));
+static int write_data(int fd, char* string,int write_len)
+{
+    ssize_t ret; 
+    while(write_len!=0)
+    {
+        ret = write(fd,string,write_len);
+        if(ret == 0)
+        {
+            break;
+        } 
+        if(ret == -1)
+        {
+            if(errno == EINTR)
+            {
+                continue;
+            }
+            syslog(LOG_ERR, "write error = %d\n",errno);
+	    fprintf(stderr, "write error: %d\n", errno);
+            return -1;
+        }
+        write_len -= ret;
+        string += ret;
+    }
+    return 0;
+}
+static int echo_file_socket(int fd, int acceptfd)
+{
+    ssize_t ret; 
+    char write_str[BUFFER_STD_SIZE];
+    while(1)
+    {
+        memset(write_str,0,sizeof(write_str));
+        ret = read(fd,write_str,sizeof(write_str));
+        if(ret == 0)
+        {
+            break;
+        } 
+        if(ret == -1)
+        {
+            if(errno == EINTR)
+            {
+                continue;
+            }
+            syslog(LOG_ERR, "read error = %d\n",errno);
+	    fprintf(stderr, "read error: %d\n", errno);
+            return -1;
+        }
+        int num_bytes_to_send = ret;
+        int num_bytes_sent = 0;
+        int str_index = 0;
+        while(num_bytes_to_send>0)
+        {
+            num_bytes_sent = send(acceptfd,&write_str[str_index],num_bytes_to_send,SEND_FLAGS);
+            if(num_bytes_sent == -1)
+            {
+		syslog(LOG_ERR, "send error = %d\n",errno);
+		fprintf(stderr, "send error: %d\n", errno);
+                return -1;
+            }
+            num_bytes_to_send -= num_bytes_sent;
+            str_index += num_bytes_sent;
+        }
+    }
+    return 0;
+}
+
+// seeked help from Guru post getting stuck
+// added goto statements
+static void* threadfunc(void* thread_param)
+{
+    char *buffer;
+    int file_fd;
+    int recv_bytes = 0,start_ptr = 0,read_bytes=0;
+    int buffer_length=0,buffer_capacity=0;
+    int status=0;
+    node_t *thread_params = (node_t *)thread_param;
+    while(1)
+    {
+        switch(thread_params->thread_type)
+        {
+            case recv_t:
+                if(buffer_capacity == buffer_length)
+                {
+                    if(buffer_capacity == 0)
+                    {
+                        buffer = malloc(BUFFER_STD_SIZE);
+                        if(buffer == NULL)
+                        {
+                            goto free_socket_fd;
```

Comment:
@abaa3023 The goto work as a cleanup strategy but are a bit more difficult to read.  You might consider splitting these logic blocks into function calls, with cleanup handled outside the function call and all necessary content passed in.
Also I'm not completely sold on the state machine design here.  I think you might be better off from a readability perspective making different functions for each state here and calling them in the desired sequence.